readthedocs/builds/models.py

```diff
@@ -698,3 +699,117 @@ def run_time(self):
+    )
+    match_arg = models.CharField(
+        _('Match argument'),
+        help_text=_('Value used for the rule to match the version'),
+        max_length=255,
+    )
+    action = models.CharField(
+        _('Action'),
+        max_length=32,
+        choices=ACTIONS,
+    )
+    action_arg = models.CharField(
+        _('Action argument'),
+        help_text=_('Value used for the action to perfom an operation'),
+        max_length=255,
+        null=True,
+        blank=True,
+    )
+    version_type = models.CharField(
+        _('Version type'),
+        max_length=32,
+        choices=VERSION_TYPES,
+    )
+
+    class Meta:
+        unique_together = (('project', 'priority'),)
+        ordering = ('priority', '-modified', '-created')
+
+    def run(self, version, *args, **kwargs):
+        """
+        Run an action if `version` matches the rule.
+
+        :type version: readthedocs.builds.models.Version
+        :returns: True if the action was performed
+        """
+        if version.type == self.version_type:
+            match_result = self.match(version, self.match_arg)
+            if match_result is not None:
+                self.apply_action(version, match_result, self.action_arg)
+                return True
+        return False
+
+    def match(self, version, match_arg):
+        """
+        Returns an object different from None if the version matches the rule.
+
+        :type version: readthedocs.builds.models.Version
+        :param str match_arg: Additional argument to perform the match
+        :returns: An object different from `None` if the match is successful.
+                  The result will be passed to `apply_action`.
+        """
+        return None
+
+    def apply_action(self, version, match_result, action_arg):
+        """
+        Apply the action from allowed_actions.
+
+        :type version: readthedocs.builds.models.Version
+        :param any match_result: Additional context from the match operation
+        :param str action_arg: Additional argument to perform the action
+        :raises: NotImplementedError if the action
+                 isn't implemented or supported for this rule.
+        """
+        action = self.allowed_actions.get(self.action)
+        if action is None:
+            raise NotImplementedError
+        action(version, match_result, action_arg)
+
+    def __str__(self):
+        class_name = self.__class__.__name__
+        return (
+            f'({self.priority}) '
+            f'{class_name}/{self.get_action_display()} '
+            f'for {self.project.slug}:{self.get_version_type_display()}'
+        )
+
+
+class RegexAutomationRule(VersionAutomationRule):
+
+    allowed_actions = {
+        VersionAutomationRule.ACTIVATE_VERSION_ACTION: actions.activate_version,
+    }
+
+    class Meta:
+        proxy = True
+
+    def match(self, version, match_arg):
+        try:
+            match = re.search(
```

Comment:
I'd say that what we want here is `fullmatch`; otherwise we are not really matching the user's regex.

`.search` produce unexpected results here:

```
In [2]: re.search('[0-9]', 'my-release-1')                                                                                                                                                    
Out[2]: <re.Match object; span=(11, 12), match='1'>

In [3]: re.fullmatch('[0-9]', 'my-release-1')                                                                                                                                                 

In [4]: re.fullmatch('[0-9]', '1')                                                                                                                                                            
Out[4]: <re.Match object; span=(0, 1), match='1'>

In [5]: re.fullmatch('[0-9]', '1a')                                                                                                                                                           

In [6]:  
```

Reply:
For me the result of fullmatch is unexpected. In regex I always need to start with `^` if I want to match from the beginning

Reply:
I think [`fullmatch`](https://docs.python.org/3.6/library/re.html#re.fullmatch) does exactly what we want: matches the full string. Using this approach will avoid false positives when matching things that are not complete, as I showed in the examples above.

Reply:
> In regex I always need to start with `^` if I want to match from the beginning

You are an advanced user :)

I'm sure that many people will end up matching versions that they don't want to match.

Reply:
The thing is that this should not be exposed to general users, so we need all the power of regex internally more than exposing this to users.